### ISSUE TITLE
Add PUT support for vector set ID

### ISSUE THREAD

**Issue by smuellerDD** [usnistgov/ACVP#757]
I know that has been discussed previously that the PUT operation on a vsID to upload a corrected response compliant to the SPEC is not yet implemented.

But may I ask to put it on the agenda to prevent more load on the production server? For example, we have obtained a large test set from the production server. We tested all ciphers before with the demo. But due to a glitch (a missing configuration setting) in the final production run, one test response failed. After correcting the glitch and having the correct configuration, the ciphers produce the correct response. However, now we need to re-download the entire large set of vsIDs as all were part of one test session. With a PUT, no additional load would be put on the production server.

Thanks

**Comment by shaneshaffer** [comment]
There are no plans to implement a PUT for a vector set, nor any other way to reattempt testing of a failed vector set. While there would be operational advantages to allowing this, there are ways that it could be abused that would compromise the integrity of the testing, so it can never be allowed.

The proper action in the event of a failed vector set is to cancel that single vector set, effectively removing it from the test session. The remainder of the test session would continue as normal. Then create a new test session, registering only the failed algorithm(s), and proceed with that test session as you would any other.

**Comment by smuellerDD** [comment]
Am Donnerstag, 16. Januar 2020, 17:31:21 CET schrieb shaneshaffer:

Hi shaneshaffer,

> There are no plans to implement a PUT for a vector set, nor any other way to
> reattempt testing of a failed vector set. While there would be operational
> advantages to allowing this, there are ways that it could be abused that
> would compromise the integrity of the testing, so it can never be allowed.
> 
> The proper action in the event of a failed vector set is to cancel that
> single vector set, effectively removing it from the test session. The
> remainder of the test session would continue as normal. Then create a new
> test session, registering only the failed algorithm(s), and proceed with
> that test session as you would any other.

How interesting. In this case, is the PUT operation now discarded? If so, 
should the ACVP spec be updated?

How would I cancel one particular vector set? Call delete on the vsID URL?

Ciao
Stephan

**Comment by AlexThurston** [comment]
I have a couple of thoughts/comments on this.  The first is that I believe the API documentation still lists PUT as a [valid route](https://htmlpreview.github.io/?https://github.com/usnistgov/ACVP/blob/master/artifacts/draft-fussell-acvp-spec-00.html#vectorSet_results_put
).  That should like be updated since it's quite misleading.

The second is that iterating on vector set responses is quite useful and I believe this should be re-visited.  While I don't quite understand how it compromises the integrity of the testing, perhaps it could be enabled on DEMO and disable on PROD as a stop gap?  This would allow for some efficiencies while working with vendors as they transition to ACVP while still preserving the integrity of testing on production?

As another solution, perhaps PUT could be enabled on test sessions that have been created as a [sample](https://htmlpreview.github.io/?https://github.com/usnistgov/ACVP/blob/master/artifacts/draft-fussell-acvp-spec-00.html#testSessions)

In this case, those test sessions aren't usable for an actual validation so integrity would be less of a concern.  This would allow production and demo to support the same API.

**Comment by AlexThurston** [comment]
An additional note to this, my second suggestion wouldn't be all that useful.  I learned in this issue https://github.com/usnistgov/ACVP/issues/780, that there is at least one case where a test session designated as a *sample* will not properly validate responses generated by the IUT.  As such, making it so that only *sample* test sessions allow PUT wouldn't help. 

However, if DEMO did and production didn't, that might be a good solution to allow for iterations on a single set of vector sets.

**Comment by AlexThurston** [comment]
As another point of interest in this discussion, I tried to follow the recommended procedure of deleting (canceling) a vector set with the failure within a test session.  I did this with a test session (73549).  I have 3 vector set within (150772, 150773, 150774).  I've uploaded passing results for the first two and a failed set for the last (150774).  I then followed @shaneshaffer recommendation of cancelling the vector set with the failed results.  However, ACVP is returning:

```json
[
  {
    "acvVersion": "1.0"
  },
  {
    "url": "/acvp/v1/testSessions/73549/vectorSets/150774",
    "status": "expired"
  }
]
```

I'm certainly still advocating to allow the PUT operation on vector sets in some form but I wanted to see how it behaved with recommended procedure.
